sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelUtils.java

```diff
@@ -51,18 +62,117 @@
+              + "Scheme: [%s] is already registered with %s, and override is not allowed.",
+          FACTORY_MAP.get(scheme).getClass(),
+          scheme,
+          factory.getClass()));
+    }
     FACTORY_MAP.put(scheme, factory);
   }
 
   /**
+   * Deregisters the scheme and the associated {@link IOChannelFactory}.
+   */
+  @VisibleForTesting
+  static void deregisterScheme(String scheme) {
+    FACTORY_MAP.remove(scheme);
+  }
+
+  /**
    * Registers standard factories globally. This requires {@link PipelineOptions}
    * to provide, e.g., credentials for GCS.
+   *
+   * @deprecated use {@link #registerIOFactories}.
    */
+  @Deprecated
   public static void registerStandardIOFactories(PipelineOptions options) {
-    setIOFactory("gs", new GcsIOChannelFactory(options.as(GcsOptions.class)));
-    setIOFactory("file", new FileIOChannelFactory());
+    registerIOFactoriesAllowOverride(options);
+  }
+
+  /**
+   * Registers all {@link IOChannelFactory IOChannelFactories} from {@link ServiceLoader}.
+   *
+   * <p>This requires {@link PipelineOptions} to provide, e.g., credentials for GCS.
+   *
+   * <p>Override existing schemes is not allowed.
+   *
+   * @throws RuntimeException if the scheme of a {@link IOChannelFactory} is already registered.
+   */
+  public static void registerIOFactories(PipelineOptions options) {
+    registerIOFactoriesInternal(options, false /* override */);
+  }
+
+  /**
+   * Registers all {@link IOChannelFactory IOChannelFactories} from {@link ServiceLoader}.
+   *
+   * <p>This requires {@link PipelineOptions} to provide, e.g., credentials for GCS.
+   *
+   * <p>Override existing schemes is allowed.
+   *
+   * @deprecated This is currently to provide different configurations for tests.
+   * Once we have configurable {@link IOChannelFactory IOChannelFactories}, we will no longer need
+   * to override. At that time, workers code load registrars once. And, rest code could
+   * create {@link IOChannelFactory IOChannelFactories} with different configurations.
+   */
+  @Deprecated
+  @VisibleForTesting
+  public static void registerIOFactoriesAllowOverride(PipelineOptions options) {
+    registerIOFactoriesInternal(options, true /* override */);
+  }
+
+  private static void registerIOFactoriesInternal(
+      PipelineOptions options, boolean override) {
+    Set<IOChannelFactoryRegistrar> registrars =
+        Sets.newTreeSet(ReflectHelpers.ObjectsClassComparator.INSTANCE);
+    registrars.addAll(Lists.newArrayList(
+        ServiceLoader.load(IOChannelFactoryRegistrar.class, CLASS_LOADER)));
+
+    checkDuplicateScheme(registrars);
+
+    for (IOChannelFactoryRegistrar registrar : registrars) {
+      setIOFactoryInternal(
+          registrar.getScheme(),
+          registrar.fromOptions(options),
+          override);
+    }
+  }
+
+  @VisibleForTesting
+  static void checkDuplicateScheme(Set<IOChannelFactoryRegistrar> registrars) {
+    Set<String> schemes = Sets.newHashSet();
```

Comment:
Use a multimap, will be easier to follow. Scheme is key, values are registrars. Will also allow you to build an error containing all the conflicting IOChannelFactoryRegistrars for a single scheme